### PR TITLE
Add S3 GetAllConcurrently function

### PR DESCRIPTION
**Pull request by CallumNZ**
Reference: https://github.com/GeoNet/tickets/issues/14787

Changes proposed in this pull request:

- Allow S3 client to be created with options.
- Add GetAllConcurrently function to download S3 objects concurrently.
- Changed a const name (which ended up not being useful)

Notes: This concurrency code is currently living in Tilde and has been used for a while. I've added memory "throttling" and updated various bits to fit in the context of a generalised S3 fan out/fan in operation (rather than Tilde specific), and added it to the S3 client with a few more configuration options.

## Production Changes

The following production changes are required to deploy these changes:

- None

## Review

Check the box that applies to this code review.  If necessary please seek help with adding a checklist guide for the reviewer.
When assigning the code review please consider the expertise needed to review the changes.

- [ ] This is a content (documentation, web page etc) only change.
- [ ] This is a minor change (meta data, bug fix, improve test coverage etc).
- [x] This is a larger change (new feature, significant refactoring etc).  Please use the code review guidelines to add a checklist below to guide the code reviewer.

### Code Review Guide

*Insert check list here if needed.*

### CODE REVIEW

**Comment by wilsonjord** [comment]
I'm likely missing something, but this seems quite a bit for `GetAll` functionality.
Is there any reason why the AWS SDK's batch downloading capability isn't used? Here is a code snippet from their documentation (I've modified it to download to buffer instead of file):

```golang
svc := s3manager.NewDownloader(session)

keys := []string{"key1", "key2", "key3"}
objectBuffers := []aws.WriteAtBuffer
objects := []s3manager.BatchDownloadObject{}

for _, key := range keys {
	buf := aws.NewWriteAtBuffer([]byte{})
	objects = append(objects, buf)

	objects = append(objects, {
		Object: &s3.GetObjectInput {
			Bucket: aws.String("bucket"),
			Key: aws.String(key),
		},
		Writer: buf,
	})

iter := &s3manager.DownloadObjectsIterator{Objects: objects}
if err := svc.DownloadWithIterator(aws.BackgroundContext(), iter); err != nil {
	return err
}

// keys, and their contents objectBuffers
```

Now I don't think it gives you parallel downloads, but I understand each individual download can be concurrent, so there's a good chance performance is still similar.

**Review by junghao — COMMENTED** [review]
.

**Comment by CallumNZ** [comment]
I chose to put it in kit so that it could be used by `strong-motion`, which does a very similar thing - download many files at once and streams to the response writer.

File: aws/s3/s3.go [hunk 1]
```diff
@@ -21,9 +21,10 @@ import (
 )
 
 type S3 struct {
-	client     *s3.Client
-	uploader   *manager.Uploader
-	downloader *manager.Downloader
+	client       *s3.Client
+	uploader     *manager.Uploader
+	downloader   *manager.Downloader
+	concurrenter *ConcurrencyManager
```
Comment (junghao) [inline aws/s3/s3.go:27]: Maybe we define a `ConcurrentS3` like 
```
type ConcurrentS3 struct {
  S3
  concurrenter *ConcurrencyManager
}
```

Then all the functions goes into `ConcurrentS3` so the current S3 remains a simple wrapping of AWS SDK S3 package?

File: aws/s3/s3.go [hunk 2]
```diff
@@ -131,6 +166,34 @@ func (s *S3) Get(bucket, key, version string, b *bytes.Buffer) error {
 	return err
 }
 
+// GetAll gets the objects specified from bucket and writes the resulting HydratedFiles
+// to the returned output channel. The closure of this channel is handled, however it's the caller's
+// responsibility to purge the channel, and handle any errors present in the HydratedFiles.
+// If the S3 client's ConcurrencyManager is not initialised before calling GetAll, an output channel
+// containing a single HydratedFile with an error is returned.
+// Version can be empty, but must be the same for all objects.
+func (s *S3) GetAll(bucket, version string, objects []types.Object) chan HydratedFile {
```
Comment (junghao) [inline aws/s3/s3.go:175]: Maybe a name like `ConcurrentGetAll` to improve the readability of the caller?

**Comment by CallumNZ** [comment]
@wilsonjord I chose not to use the Batch Downloader because from what I read it seemed the order that each object comes back is not guaranteed, but I haven't tried it to see if that's the case. The output order is the biggest advantage of this implementation, since it's much lighter to sort keys beforehand, and each object can be written straight to the response writer, at which point it can be garbage collected.

**Comment by wilsonjord** [comment]
> The output order is the biggest advantage of this implementation, since it's much lighter to sort keys beforehand, and each object can be written straight to the response writer, at which point it can be garbage collected.

The output order in the code snippet is defined by whatever you like; in this case, the order of the output (`objectBuffers`) is the same as `keys`. Is this what you mean by ordered?

**Comment by wilsonjord** [comment]
Just extra context around the code snippet: it is based off `Downloader.DownloadWithIterator`: https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#Downloader.DownloadWithIterator

Regarding buffers used in the code snippet, I believe the struct definition implies dynamic growth, from the docs:
```golang
type WriteAtBuffer struct {

    // GrowthCoeff defines the growth rate of the internal buffer. By
    // default, the growth rate is 1, where expanding the internal
    // buffer will allocate only enough capacity to fit the new expected
    // length.
    GrowthCoeff float64
    // contains filtered or unexported fields
}
```
https://docs.aws.amazon.com/sdk-for-go/api/aws/#NewWriteAtBuffer

**Comment by CallumNZ** [comment]
https://github.com/GeoNet/tilde/pull/225/files

Line 55 of `file.go` is where the S3 client with concurrency manager is setup up, and the call is on line 235 of `observation.go`

**Review by junghao — APPROVED** [review]
Really good.

**Pull request merged**